### PR TITLE
fix(agent): enforce shadow account expiry on ShellHub logins

### PR DESCRIPTION
**Pull request by otavio**
## What
The agent now enforces account expiry and password aging from the host's account files on SSH
logins made through ShellHub. On FreeBSD, an empty password hash only logs in when
`SHELLHUB_PERMIT_EMPTY_PASSWORDS` is set.

## Why
The agent checked the password hash in `/etc/shadow` and ignored the expiry and aging fields.
Nothing replaced the PAM account check it lost when it dropped CGO. An account an administrator
had expired still logged in through ShellHub with its password, while local sshd refused it. On
FreeBSD, an account with an empty hash logged in with an empty password, and no setting turned
that off.

## Changes
- **Password login (Linux)**: follows Linux-PAM's `check_shadow_expiry`. The account is refused
  on and after its expire day. The password is refused once `today - lastchg >= max`, or when
  `lastchg` is 0. An empty or negative field disables its check. Where local sshd would force a
  password change, the agent refuses the login: it has no way to run that change.
- **Password login (FreeBSD)**: reads `change` and `expire` from `master.passwd`, which are epoch
  seconds with 0 meaning unset, and refuses at `now >=` either one, as FreeBSD's pam_unix does.
- **Public-key login**: sshd runs the PAM account check for keys too, so host mode
  (`Backend.AccountExpired`) and connector mode (the container's `/etc/shadow`) now refuse expired
  accounts. An aged-out password does not block a key.
    - A non-root agent can't read the shadow file and runs as a single user, so it skips the check.
    - A container with no `/etc/shadow` has no expiry to enforce.
    - Any other read failure refuses the login.
- **Parsing**: a malformed number in a shadow line used to be read as 0. That line is now skipped
  and logged by line number, as glibc does, so a bad line can't lock out every account on the host.
- **FreeBSD empty hash** (separate commit): matches the Linux behavior, including accepting any
  password when the flag is on, as pam_unix's `nullok` does.

## Testing
- Aging tests mock `pkg/clock` and call `AuthUserFromShadow`, so the epoch-to-days conversion is
  tested too. The boundary cases are the expire day itself and `today - lastchg == max`, both refused.
- Upgrade risk to check: a host whose accounts have `lastchg` set to 0 (a forced password change),
  or whose passwords are past `max`, loses password login through ShellHub. Key login still works.
- The FreeBSD tests only build on FreeBSD, so CI doesn't run them. They were compiled and vetted
  with `GOOS=freebsd`.
- The connector-mode key check has no unit test, because the package has no Docker client test
  setup.

### CODE REVIEW

File: agent/pkg/osauth/auth.go [hunk 1]
```diff
@@ -42,6 +45,26 @@ func (b *backend) AuthUser(username, password string) bool {
...
+	if errors.Is(err, fs.ErrNotExist) {
+		return false
+	}
+
+	if err != nil {
+		logrus.WithError(err).Error("Error opening shadow file")
+
+		return true
+	}
+	defer file.Close() //nolint:errcheck
```
[the remark this answers is not in the document]
Comment (otavio) [reply agent/pkg/osauth/auth.go]: Addressed in e51b8c0d9.

File: agent/pkg/osauth/auth.go [hunk 2]
```diff
@@ -95,7 +170,46 @@ func AuthUserFromShadow(username, password string, shadow io.Reader) bool {
...
+	entries, err := parseShadowReader(shadow)
+	if err != nil {
+		logrus.WithError(err).Error("Error parsing shadow file")
+
+		return true
+	}
+
+	entry, ok := entries[username]
+	if !ok {
+		return false
```
[the remark this answers is not in the document]
Comment (otavio) [reply agent/pkg/osauth/auth.go]: Addressed in e51b8c0d9: any unparsable line now refuses only its own account, on both password and key paths (FreeBSD's malformed change/expire too).

File: agent/server/modes/connector/authenticator.go [hunk 3]
```diff
@@ -68,6 +69,26 @@ func getShadow(ctx context.Context, cli dockerclient.APIClient, container string
 	return shadow, nil
 }
 
+func accountExpiredInContainer(ctx context.Context, cli dockerclient.APIClient, container, username string) bool {
+	shadow, err := getShadow(ctx, cli, container)
```
[the remark this answers is not in the document]
Comment (otavio) [reply agent/server/modes/connector/authenticator.go:74]: Addressed in e51b8c0d9: the passwd/shadow readers now carry the response body as their closer and every caller defers it.

File: agent/server/modes/connector/authenticator.go [hunk 4]
```diff
@@ -68,6 +69,26 @@ func getShadow(ctx context.Context, cli dockerclient.APIClient, container string
...
+	if cerrdefs.IsNotFound(err) {
+		return false
+	}
```
[the remark this answers is not in the document]
Comment (otavio) [reply agent/server/modes/connector/authenticator.go:88]: Addressed in e51b8c0d9: on NotFound the container is inspected, and the login is refused unless it still exists.

File: agent/pkg/osauth/auth.go [hunk 5]
```diff
@@ -67,6 +90,49 @@ type shadowEntry struct {
...
+	case e.Lastchanged == 0:
+		return errPasswordChangeRequired
+	case e.Lastchanged < 0, e.Maximum < 0, today < e.Lastchanged:
+		return nil
+	case today-e.Lastchanged >= e.Maximum:
+		return errPasswordExpired
+	default:
+		return nil
+	}
+}
```
[the remark this answers is not in the document]
Comment (otavio) [reply agent/pkg/osauth/auth.go:127]: Won't address: current Linux-PAM check_shadow_expiry has `if (spent->sp_lstchg < 0) return PAM_SUCCESS` and `passed >= spent->sp_max`, so both behaviours already match PAM.

File: agent/pkg/osauth/auth.go [hunk 6]
```diff
@@ -67,6 +90,49 @@ type shadowEntry struct {
...
+
+func (e shadowEntry) checkPasswordLogin(today int) error {
+	if err := e.checkAccountExpiry(today); err != nil {
+		return err
+	}
+
+	return e.checkPasswordAge(today)
+}
```
[the remark this answers is not in the document]
Comment (otavio) [reply agent/pkg/osauth/auth.go:135]: Won't address: lifting the policy means parameterising two unit systems and sentinels for ~20 lines; the real gap is FreeBSD tests not running in CI, which is wider than this PR.
